File: 2855-maximum-number-of-jumps-to-reach-the-last-index/maximum-number-of-jumps-to-reach-the-last-index.java
class Solution {
    public int maximumJumps(int[] nums, int target) {
        int n = nums.length;
        
        // dp[i] = maximum jumps needed to reach index i
        int[] dp = new int[n];
        
        // initialize as impossible
        for (int i = 0; i < n; i++) {
            dp[i] = -1;
        }
        
        dp[0] = 0; // starting point
        
        for (int i = 0; i < n; i++) {
            if (dp[i] == -1) continue;
            
            for (int j = i + 1; j < n; j++) {
                if (Math.abs(nums[j] - nums[i]) <= target) {
                    dp[j] = Math.max(dp[j], dp[i] + 1);
                }
            }
        }
        
        return dp[n - 1];
    }
}

File: 2855-maximum-number-of-jumps-to-reach-the-last-index/maximum-number-of-jumps-to-reach-the-last-index.cpp
class Solution {
public:
    int maximumJumps(vector<int>& nums, int target) {
        int n = nums.size();

        vector<int> dp(n, -1);

        // We start at index 0
        dp[0] = 0;

        for (int i = 0; i < n; i++) {

            // If index i cannot be reached, skip it
            if (dp[i] == -1)
                continue;

            // Try jumping from i to every index after it
            for (int j = i + 1; j < n; j++) {

                if (abs(nums[j] - nums[i]) <= target) {

                    dp[j] = max(dp[j], dp[i] + 1);
                }
            }
        }

        return dp[n - 1];
    }
};